### PR TITLE
Do not do redirect handling when loggin out

### LINKED ISSUE #12568: Logout keeps old URL as redirect URL in Firefox?
@schiessle and me experience this problem in Firefox. Neither Chromium nor Safari suffers from this issue, but we have not idea what's causing it.

1. As an admin log in
2. Go to admin settings
3. Click log out
4. URL is now `index.php/login?redirect_url=/index.php/settings/admin/overview`
5. Log in as a normal user
6. :boom: 
> **Access forbidden**
> Logged in user must be an admin

Actually the mistake is in step 4. After the logout you should not have the redirect_url anymore. Happens in latest 14 + master, not sure when it started.

### PR DESCRIPTION
**Pull request by rullzer**
Fixes #12568
Since the clearing of the execution context causes another reload. We
should not do the redirect_uri handling as this results in redirecting
back to the logout page on login.

Signed-off-by: Roeland Jago Douma <roeland@famdouma.nl>

### CODE REVIEW

**Comment by nickvergessen** [comment]
I dislike this patch a bit, because if you are on a page which is publicly accessible, it loads that page, instead of redirecting to the login form.

In my head the following should have worked:
```php
	public function showLoginForm(string $user = null, string $redirect_url = null): Http\Response {
		if ($this->session->exists('clearingExecutionContexts')) {
			$this->session->remove('clearingExecutionContexts');

			$response = new RedirectResponse($this->urlGenerator->linkToRouteAbsolute('core.login.showLoginForm'));
			$response->addHeader('Clear-Site-Data', '"cache", "storage", "executionContexts"');
			return $response;
		}
```

But while your request is on `/login` firefox reloads the last loaded page, which is still the previous page, so it also would reload the public call.

**Comment by rullzer** [comment]
Do you have a logout button on a public page?

**Comment by nickvergessen** [comment]
Well on Talk the same URL is used for a room independent of your login state.

**Comment by MorrisJobke** [comment]
@rullzer @nickvergessen What is the status here? Continue it or close it?

**Comment by rakekniven** [comment]
No, it is still the same with ff 65.

**Comment by nickvergessen** [comment]
Same here, still broken with FF 65.0

**Comment by rullzer** [comment]
Ok this is now a bit more elegant I think.

@nickvergessen please check it out

File: lib/private/AppFramework/DependencyInjection/DIContainer.php [hunk 1]
```diff
@@ -185,6 +185,13 @@ public function __construct($appName, $urlParams = array(), ServerContainer $ser
 			$server =  $this->getServer();
 
 			$dispatcher = new MiddlewareDispatcher();
+			$dispatcher->registerMiddleware(
+				new OC\AppFramework\Middleware\Security\ReloadExecutionMiddleware(
+					$c->query(ISession::class),
+					$c->query(IURLGenerator::class)
+				)
+			);
```
Comment (MorrisJobke) [inline lib/private/AppFramework/DependencyInjection/DIContainer.php]: Out of curiosity: Isn't it possible to just query the middleware? 🤔

File: lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php [hunk 2]
```diff
@@ -52,6 +52,7 @@
 use OCP\AppFramework\OCSController;
 use OCP\IL10N;
 use OCP\INavigationManager;
+use OCP\ISession;
```
Comment (MorrisJobke) [inline lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php]: I guess this is not needed anymore.

File: lib/private/AppFramework/DependencyInjection/DIContainer.php (hunk 1 above)
Comment (rullzer) [reply lib/private/AppFramework/DependencyInjection/DIContainer.php]: Of course it is

File: lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php (hunk 2 above)
Comment (rullzer) [reply lib/private/AppFramework/Middleware/Security/SecurityMiddleware.php]: Yep.

**Review by nickvergessen — APPROVED** [review]
That works perfectly.

**Review by MorrisJobke — APPROVED** [review]
Tested and works 👍

**Comment by MorrisJobke** [comment]
Backport?

**Comment by rullzer** [comment]
Fine by me. It is mainly an added middleware. So should not cause issues.

**Pull request merged**